tests/cpp/multidevice.cpp

```diff
@@ -25,6 +28,12 @@ MultiDeviceTest::MultiDeviceTest() {
       (getNvFuserEnv("MULTIDEVICE_DEBUG_BARRIER") != nullptr &&
        communicator->is_available());
   disable_skip = getNvFuserEnv("MULTIDEVICE_DISABLE_SKIP") != nullptr;
+
+  // Currently, we attempt to wait for debugger before every test, which is
+  // fine because we usually debug one test at a time using --gtest_filter. As
+  // a potential improvement, call it at the test program level:
+  // https://github.com/google/googletest/blob/main/docs/advanced.md#global-set-up-and-tear-down
```

Comment:
Thank you for calling this out with a comment.

This seems like it would be better, in particular because we already have the `do_barrier_at_test` variable. Presumably somebody added that variable because sometimes we want that barrier and sometimes we don't.

But a difficulty is that one can't `waitForDebugger` until the process has gotten through `MPI_Init`. Is that guaranteed the SetUpTestSuite time?

Anyway, I'd like to see one or both of:
* If we want to go this route, some comments / a discussion (with @samnordmann and @cowanmeg) about whether we should remove the `do_barrier_at_test` logic (since this effectively forces it on).
* A switch to the SetUpTestSuite once-per-suite and not once-per-test approach referenced in the comment

Reply:
Great suggestions -- thank you! 

> since this effectively forces it on

To be more accurate, this effectively forces it on when NVFUSER_MPI_DEBUG_RANK is specified; otherwise, waitForDebugger returns immediately. IIUC, for debugging, it's better to turn on the barrier anyway. 

> A switch to the SetUpTestSuite once-per-suite and not once-per-test approach referenced in the comment

Let me look into how difficult this is.

Reply:
> This seems like it would be better, in particular because we already have the `do_barrier_at_test` variable. Presumably somebody added that variable because sometimes we want that barrier and sometimes we don't.

> If we want to go this route, some comments / a discussion (with @samnordmann and @cowanmeg) about whether we should remove the do_barrier_at_test logic (since this effectively forces it on).

Why is it a problem to combine the barrier and the present feature?

By the way, I recall that the barrier is non-blocking from the host perspective. Nccl implements it as an Allreduce.

> But a difficulty is that one can't waitForDebugger until the process has gotten through MPI_Init. Is that guaranteed by SetUpTestSuite time?

There is no `MPI_Init` in the program. However I don't think this is a problem: when `waitForDebugger` is called, the `communicator` is already instantiated (through `getOrCreateCommunicator`), so the backend is already available and we can post a barrier.

> > A switch to the SetUpTestSuite once-per-suite and not once-per-test approach referenced in the comment

> Let me look into how difficult this is.

Idea: this could probably be achieved with gtest's environment (that has been removed recently https://github.com/NVIDIA/Fuser/pull/2117), during environment's setup.

Reply:
> Great suggestions -- thank you!
> 
> > since this effectively forces it on
> 
> To be more accurate, this effectively forces it on when NVFUSER_MPI_DEBUG_RANK is specified; otherwise, waitForDebugger returns immediately. IIUC, for debugging, it's better to turn on the barrier anyway.

Sorry, the "it" I used was indeed ambiguous :-)

I meant that this basically forces `do_barrier_at_test` on. But I see now that I misread, and it never reaches the `communication->barrier()` when the env var is unset, because of an early exit.

> Why is it a problem to combine the barrier and the present feature?

It's not, given my misread of the code.

Reply:
I changed the code to call waitForDebugger only once by using std::call_once.